contracts/JBTieredNFTRewardDelegate.sol

```diff
@@ -0,0 +1,381 @@
+        _tier.idCeiling > _tokenId && _tier.idCeiling - _tier.initialAllowance <= _tokenId
+        // 1 indexed, meaning the 0th element of the array is tier 1.
+      ) return _i + 1;
+
+      unchecked {
+        ++_i;
+      }
+    }
+
+    return 0;
+  }
+
+  //*********************************************************************//
+  // -------------------------- constructor ---------------------------- //
+  //*********************************************************************//
+
+  /**
+    @param _projectId The ID of the project for which this NFT should be minted in response to payments made. 
+    @param _directory The directory of terminals and controllers for projects.
+    @param _name The name of the token.
+    @param _symbol The symbol that the token should be represented by.
+    @param _tokenUriResolver A contract responsible for resolving the token URI for each token ID.
+    @param _baseUri The token's base URI, to be used if a URI resolver is not provided. 
+    @param _contractUri A URI where contract metadata can be found. 
+    @param _expectedCaller The address that should be calling calling the data source.
+    @param _owner The address that should own this contract.
+    @param _contributionToken The token that contributions are expected to be in terms of.
+    @param __tiers The tiers according to which token distribution will be made. 
+  */
+  constructor(
+    uint256 _projectId,
+    IJBDirectory _directory,
+    string memory _name,
+    string memory _symbol,
+    IToken721UriResolver _tokenUriResolver,
+    string memory _baseUri,
+    string memory _contractUri,
+    address _expectedCaller,
+    address _owner,
+    address _contributionToken,
+    JBNFTRewardTier[] memory __tiers
+  )
+    JBNFTRewardDataSource(
+      _projectId,
+      _directory,
+      _name,
+      _symbol,
+      _tokenUriResolver,
+      _baseUri,
+      _contractUri,
+      _expectedCaller,
+      _owner
+    )
+  {
+    contributionToken = _contributionToken;
+
+    // Make sure the tiers were delivered in order.
+    if (__tiers.length != 0) {
+      _tiers.push(__tiers[0]);
+
+      // Get a reference to the number of tiers.
+      uint256 _numTiers = __tiers.length;
+
+      // Keep a reference to the tier being iterated on.
+      JBNFTRewardTier memory _tier;
+
+      for (uint256 _i = 1; _i < _numTiers; ) {
+        // Set the tier being iterated on.
+        _tier = __tiers[_i];
+
+        // Make sure the tier's contribution floor is greater than the previous contribution floor.
+        if (_tier.contributionFloor <= __tiers[_i - 1].contributionFloor)
+          revert INVALID_PRICE_SORT_ORDER(_i);
+
+        // Make sure the tiers' ID ranges don't collide.
+        if (_tier.idCeiling - _tier.remainingAllowance < __tiers[_i - 1].idCeiling)
+          revert INVALID_ID_SORT_ORDER(_i);
+
+        // Set the initial allowance to be the remaining allowance.
+        _tier.initialAllowance = _tier.remainingAllowance;
```

Comment:
This doesn't work for the first tier (which is pushed before this loop)

Reply:
good catch